File: packages/ts/audit/src/audit.ts
import type { AuditEntry, AuditLogger } from "@drakkar.software/starfish-protocol"

/** Audit logger that writes to console. */
export function createConsoleAuditLogger(): AuditLogger {
  return {
    record(entry) {
      const status = entry.success ? "OK" : "FAIL"
      console.log(
        `[Starfish:AUDIT] ${entry.action.toUpperCase()} ${entry.collection} ` +
        `by ${entry.identity ?? "anonymous"} → ${status} (${entry.statusCode})`,
      )
    },
  }
}

/** Audit logger that delegates to a callback. */
export function createCallbackAuditLogger(
  cb: (entry: AuditEntry) => void | Promise<void>,
): AuditLogger {
  return { record: cb }
}

/** No-op audit logger (discards entries). */
export function createNoopAuditLogger(): AuditLogger {
  return { record() {} }
}

File: packages/python/audit/starfish_audit/audit.py
"""Concrete audit loggers for the Starfish server.

The ``AuditEntry`` dataclass and ``AuditLogger`` base class live in
``starfish_protocol``; this module supplies the ready-made implementations.
"""

import inspect
from typing import Awaitable, Callable

from starfish_protocol import AuditEntry, AuditLogger


class ConsoleAuditLogger(AuditLogger):
    """Audit logger that writes to console."""

    async def record(self, entry: AuditEntry) -> None:
        status = "OK" if entry.success else "FAIL"
        identity = entry.identity or "anonymous"
        print(
            f"[Starfish:AUDIT] {entry.action.upper()} {entry.collection} "
            f"by {identity} → {status} ({entry.status_code})"
        )


class CallbackAuditLogger(AuditLogger):
    """Audit logger that delegates to a sync or async callback."""

    def __init__(self, callback: Callable[[AuditEntry], None | Awaitable[None]]) -> None:
        self._callback = callback

    async def record(self, entry: AuditEntry) -> None:
        result = self._callback(entry)
        if inspect.isawaitable(result):
            await result


class NoopAuditLogger(AuditLogger):
    """No-op audit logger (discards entries)."""

    async def record(self, entry: AuditEntry) -> None:
        pass


__all__ = ["ConsoleAuditLogger", "CallbackAuditLogger", "NoopAuditLogger"]
